pyface/base_application.py

```diff
@@ -0,0 +1,395 @@
+        return True
+
+    def stop(self):
+        """ Stop the application, cleanly releasing resources if possible. """
+        logger.info('---- Application stopping ----')
+        return True
+
+    def run(self):
+        """ Run the application
+
+        Return
+        ------
+        status : bool
+            Whether or not the application ran normally
+        """
+        started = run = stopped = False
+
+        # Start up the application.
+        self._fire_application_event('starting')
+        started = self.start()
+        if started:
+
+            logger.info('---- Application started ----')
+            self._fire_application_event('started')
+
+            try:
+                run = self._run()
+            except ApplicationExit as exc:
+                if exc.args == ():
+                    logger.info("---- ApplicationExit raised ----")
+                else:
+                    logger.exception("---- ApplicationExit raised ----")
+                run = (exc.args == ())
+            finally:
+                # Try to shut the application down.
+                self._fire_application_event('stopping')
+                stopped = self.stop()
+                if stopped:
+                    self._fire_application_event('stopped')
+                    logger.info('---- Application stopped ----')
+
+        return started and run and stopped
+
+    def exit(self, force=False):
+        """ Exits the application.
+
+        This method handles a request to shut down the application by the user,
+        eg. from a menu.  If force is False, the application can potentially
+        veto the close event, leaving the application in the state that it was
+        before the exit method was called.
+
+        Parameters
+        ----------
+        force : bool, optional (default False)
+            If set, windows will receive no closing events and will be
+            destroyed unconditionally. This can be useful for reliably tearing
+            down regression tests, but should be used with caution.
+
+        Raises
+        ------
+        ApplicationExit
+            Some subclasses may trigger the exit by raising ApplicationExit.
+        """
+        logger.info('---- Application exit started ----')
+        if force or self._can_exit():
+            try:
+                self._prepare_exit()
+            except Exception:
+                logger.exception("Error preparing for application exit")
+            finally:
+                logger.info('---- Application exit ----')
+                self._exit()
+        else:
+            logger.info('---- Application exit vetoed ----')
+
+    # Initialization utilities -----------------------------------------------
+
+    def setup_logging(self):
+        """ Initialize logger.
+
+        Default behaviour is to log to a rotating file handler whose name is
+        the id of the app.
+        """
+        root_logger = logging.getLogger()
+
+        if DEBUG:
```

Comment:
What about making this an instance attribute (or an argument to the method) instead of a module level variable? Would make it easier to access for end users.